src/framework/auth/services/auth.service.ts

```diff
@@ -34,14 +34,40 @@ export class NbAuthService {
   }
 
   /**
-   * Returns true if auth token is presented in the token storage
-   * @returns {Observable<any>}
+   * Returns true if auth token is present in the token storage
+   * @returns {Observable<boolean>}
    */
   isAuthenticated(): Observable<boolean> {
     return this.getToken()
       .pipe(map((token: NbAuthToken) => token.isValid()));
   }
 
+  /**
+   * Returns true if valid auth token is present in the token storage.
+   * If not, calls the strategy refreshToken, and returns isAuthenticated() if success, false otherwise
+   * @returns {Observable<boolean>}
+   */
+  isAuthenticatedOrRefresh(): Observable<boolean> {
+    return this.getToken()
+      .pipe(
+        switchMap(token => {
+        if (!token.isValid()) {
+          return this.refreshToken(token.getOwnerStrategyName(), token)
```

Comment:
Does this work with any kind of token type? Is the refreshToken endpoint always expecting the same token that was stored?
I am using an API that during authentication returns an object consisting of a refresh token, an access token and other properties. However, on refreshing the token I am expected to only supply the refresh token.

To get it working, I had to convert the token to an NbAuthOAuth2Token and explicitly call "oAuth2TOken.getRefreshToken()" on this line instead of just "token".

Reply:
Hi @cloakedch ,

The refresh_token is implemented according to [RFC6749 section 1.5](https://tools.ietf.org/html/rfc6749#section-1.5).

This RFC specifies the OAuth2 authentication scheme. So the code works with auths server returning tokens according to the [RFC section 4.2](https://tools.ietf.org/html/rfc6749#section-4.2.2)
The refreshToken endpoint is waiting for the refresh_token that was sent during login or during the latest refresh_token call. Both `grant_type=refresh_token` and `refresh_token=theRefreshToken`  are sent to the refreshToken endpoint for getting a new access-token, according to [the RFC](https://tools.ietf.org/html/rfc6749#page-47).

I would say that you  have to use the `NbOAuth2AuthStrategy`  with `password` grantType(if you send login/password at login) and `NbAuthOAuth2Token`class in your case.

Someting like
```
            NbOAuth2AuthStrategy.setup({
          name: 'password',
          clientId: 'test',
          clientSecret: 'secret',
          baseEndpoint: 'http://your.BaseEnd/Point',
          token: {
            endpoint: 'token',
            grantType: NbOAuth2GrantType.PASSWORD,
            class: NbAuthOAuth2Token,
          },
          refresh: {
            endpoint: 'refresh-token',
            grantType: NbOAuth2GrantType.REFRESH_TOKEN,
          },
        }),
```
Hope it helps

Reply:
OK, I guess the API I am using is then not working according to the RFC.
Thanks for clearing that up!